### ISSUE TITLE
Some icons need improvement for dark themes

### ISSUE THREAD

**Issue by dliessi** [wbsoft/frescobaldi#642]
Reported by @proski in https://github.com/wbsoft/frescobaldi/issues/637#issuecomment-78126940.

**Comment by proski** [comment]
Ubuntu (the default Unity desktop) uses a dark theme. Following icons in the menu are hard to see:
Snippets/Repeat
Snippets/Last note or chord
The corresponding icons are `bar_repeat_start.svg` and `note_4d.svg`

**Comment by wbsoft** [comment]
Both icons reside in `symbols` and are used e.g. on the buttons of the quick insert panel. They should automatically be drawn in the default text color. See `symbols/__init__.py`

**Comment by wbsoft** [comment]
The icons should automatically redraw when chaning the desktop theme. But it seems that some widgets do not redraw immediately, or only after the first change in desktop theme. On my system (Xubuntu), after restarting Frescobaldi after changing the desktop theme, the icons are redrawn in the text color.

**Comment by PeterBjuhr** [comment]
I did a quick test. The color changer in `pixmap` does work. The problem seems to be that the menu text color, which is basically white in this case, isn't correctly transferred.

**Comment by PeterBjuhr** [comment]
The color is set by this line in `symbols.pixmap`:

    color = QApplication.palette().foreground().color()

Should it be the widget palette instead of the application palette??

**Comment by PeterBjuhr** [comment]
I checked with the menu item directly and its palette. The result was the same. So I guess the desktop theme doesn't change the palette at all.

**Comment by wbsoft** [comment]
It might be a Qt issue. Often changing the theme colors work one time perfectly, but subsequent changes are not completely performed, i.e. some parts of Frescobaldi remain in the wrong color. After a restart all is well.

**Comment by PeterBjuhr** [comment]
We could add a solid white background to the icons. But having seen the change to white lines in practice I think that is much nicer.

**Comment by PeterBjuhr** [comment]
Finally I've found something useful. In Ubuntu at least the menu text color seems to be defined as `toolTipText`.

````python
color = QApplication.palette().toolTipText().color()
````

The above does work when I switch between the default unity theme (ambiance) with dark menus and others with lighter menus.

Two questions remain:

1. Is this something specific for Qt on Ubuntu or is this generic for all OS:s? Here I think I need help testing!

2. Can we detect that the icon is in a menu and use the above line only for those cases? I guess it would be possible but haven't researched this part yet.

**Comment by proski** [comment]
Not quite good. White (rather than black) on light gray for inactive entries. Lubuntu 14.10, Lubuntu-default theme.

I really don't want to make any decisions based on experiments with a few themes. Qt uses certain colors to draw the menu text. Those colors should be available to Qt programs. If they are not, it's a Qt bug. If Qt programs can get the colors, but PyQt programs cannot, it's a PyQt bug. If there are such limitations, let's report them and see if there are any suggestions from Qt/PyQt developers.

Then we can make guesses based on the existing relevant information. Tooltip colors are irrelevant. Inactive menu color is relevant. If it's bright, we can use black for symbols, otherwise we can use white. It's more important that inactive icons are visible. They allow users to find the right entry quickly.

![tooltip-test](https://cloud.githubusercontent.com/assets/1317472/6648311/c2cce7a2-c9ae-11e4-83ac-1210a8822004.png)

**Comment by wbsoft** [comment]
The tooltip color issue is a Qt bug: https://bugs.launchpad.net/ubuntu/+source/qt4-x11/+bug/877236

We could try and get the foreground color of the widget that displays the icon (be it a toolbutton in the quick insert panel or a menu in the main menu). But on my systems that works already. Only it seems that when changing the color multiple times, it is not picked up by all widgets (sometimes).

**Comment by PeterBjuhr** [comment]
@proski thanks for the testing! The tooltip idea failed on first try. But it was obviously unlikely. 

The point about active vs inactive menu colors is interesting. It seems to me that with the current method of obtaining a color from the palette we can never accommodate for these different states.

I had another idea of somehow using `QApplication.style()` to obtain the relevant color information. But haven't found a way yet.

**Comment by PeterBjuhr** [comment]
@wbsoft  I can confirm the tooltip bug. `toolTipBase().color()` (the tooltip background) returns a light color instead of a dark.

We need to somehow get the information from the widget that displays the icon which `QPalette::ColorRole` (WindowText, ToolTipText etc) and which `QPalette::ColorGroup` (Active, Inactive, disabled) it's using.

Another possible way forward (as mentioned above) is that the `QStyle` holds all information needed and we can use that to recolor the icons.

**Comment by proski** [comment]
I asked on StackOverflow - no replies so far.
http://stackoverflow.com/questions/29045905/pyqt-how-to-find-menu-text-color-in-that-works-with-ubuntu-unity
The Ubuntu Unity menu is called Application Menu or Appmenu. The Qt implementation of Appmenu is available here:
https://launchpad.net/appmenu-qt
The sources have some interesting stuff inside. It is possible to disable Appmenu either with environment variables or with a `setNativeMenuBar()` call. The menubar type can be requested by `isNativeMenuBar()`. Those functions are available in PyQt. So we have a few fallback solutions.

- Disable Appmenu in Frescobaldi.
- Render the symbol background if `mainwindow.menuBar().isNativeMenuBar()` is true.
- Do some guesswork/tooltip magic if `mainwindow.menuBar().isNativeMenuBar()` is true.
- Use a color visible on white and dark gray, e.g. red.

I still hope we find a way to determine the menu text color. Or maybe the Icon Engine can do some magic.

**Comment by wbsoft** [comment]
Indeed we should try to get the color in a smart way. As a fallback we could perform some graphic manipulation to the pixmap on request.

**Comment by PeterBjuhr** [comment]
About active vs inactive menu colors, `symbols.pixmap` also has the parameters `mode` and `state`.

When the menu item with the symbol gets active it doesn't change the mode unfortunately. (But when a snippet with a symbol gets selected the mode does change.)

**Comment by PeterBjuhr** [comment]
When `mode` is set to selected it seems pretty straight-forward to change the color accordingly:

````python
if mode == QIcon.Selected:
    color = QApplication.palette().highlightedText().color()
else:
    color = QApplication.palette().foreground().color()
````

**Comment by proski** [comment]
Good news! The palette of the menubar has the right color. The patch is here:
https://github.com/proski/frescobaldi/commit/66c70866e23a195d8cba9644b4ebd101499f959d
I gave it too little testing to submit a pull request.
@PeterBjuhr I'll have a look whether your patch needs to be incorporated. It can be easy combined with my patch.

**Comment by PeterBjuhr** [comment]
Great news! Your patch looks good at first glance!

My mode check isn't important, but it could be nice that the symbol changes color as well when the snippet is selected.

**Comment by wbsoft** [comment]
I will look if I can simplify the implementation by not using the palette of a widget but some QStyle constant.

**Comment by PeterBjuhr** [comment]
>>`key = (name, size.width(), size.height(), color.rgb(), mode, palette)`

>I see a QPalette is hashable, so this looks good.

Do we need the palette in the key? As long as the color is the same we don't need to know which palette it came from.

**Comment by wbsoft** [comment]
Note that on creation of the icon, it is not known where it will displayed. Also, when the user changes the desktop color theme, the icons are not created again. (Currently, the Snippets menu is redrawn on every display, but this might change in the future.)

in the `paint` method of QIconEngineV2 we get the painter, and `painter.paintEngine().paintDevice()` returns the QPaintDevice the icon is to be drawn upon. Casting that to a QWidget works, but calling any method on the widgets crashes Python, both 2 and 3.

We could draw a background or a blurred shadow below the icon, so that it shows on any background.

We could also refrain from using symbols in the menus etc. They were originally designed for the buttons in the Quick Insert panel.

**Comment by proski** [comment]
@PeterBjuhr QPalette is not hashable in Python 3. I simply didn't have a change to test it. It means we need another solution. We need the highlighted color. Just because the default Ubuntu theme uses the same color, it doesn't mean other themes would do that. That means having two colors in the key. As I understand, one QIcon can have several pixmaps (up to 8).

**Comment by proski** [comment]
@wbsoft Maybe the icon cache could be emptied when the the theme changes? I'm OK if the symbols are shown incorrectly after the theme change. It's not a big deal to restart Frescobaldi. I'm more concerned that Frescobaldi shows barely visible icons in the default theme.

**Comment by proski** [comment]
I tested other themes on Ubuntu (I'm using version 12.04 for my testing). Unfortunately, the High Contrast theme needs a highlight color, but the corresponding pixmap is never requested. It means that the line under the mouse cursor has a black icon on black background. As I said before, we can live with it (non-highlighted entries are more important), but it's an imperfection that would probably need to be accepted. The appmenu operates separately from the applications and talks to the applications over DBus. I don't think we can force-feed the active icon pixmap to the appmenu. There is no QIcon constructor that would take more than one pixmap. I only hope that appmenu supports highlighted icons on newer version of Ubuntu. So I'm going to publish a new version of my patch with highlight support, whether it works or not.

**Comment by proski** [comment]
e27313b

Support for `highlightedText` - thanks to @PeterBjuhr . Confirmed not working on Ubuntu 12.04, still hoping it might work somewhere.

Python 3 compatibility fix - color rgba values are used in key.

Replaced `rgb()` with `rgba()` in another key for consistency and safety.

Key calculation moved outside the try block.

`icon()` in `symbols/__init__.py` accepts `None` as palette, caller simplified.

**Comment by proski** [comment]
@wbsoft The caller knows how it would use the icon. By the way, if the icon is made for using in a menu, perhaps it should use the `text()` color rather than `foreground()`.

Since the icons are sent to the appmenu over dbus, it's probably not important what widget they are painted on. I believe the reason why the menubar palette helps is because it's a palette used by the menubar when the appmenu is disabled. The theme engine sets the same palette to the appmenu and the normally hidden in-window menubar. Using the QMenu palette doesn't work, I tried that. So it won't help us to find the widget using the icon.

However, if the crash prevents you from implementing some cool idea, please post the code. Maybe it can be fixed.

We could pre-generate icons from symbols by adding a background or a border. But symbols matching the text color in the menu look great (most of the time), it would be a pity to lose them.

**Comment by wbsoft** [comment]
When a QIcon is added to a QAction, it can be used on buttons, menus and in the keyboard shortcuts list in the prefs, everywhere; you can't know beforehand. When the icon is painted by the `QIconEngineV2` its `pixmap()` method is called to get a pixmap. The default implementation of `QIconEngineV2.pixmap()` calls `paint()` to paint the icon, and then we could try to know what device the painter is painting on. I overloaded `paint()` to return the pixmap, which is painted separately and then cached. It would be very nice if we could get palette information there... in the `paint()` method of `QIconEngineV2`

**Comment by PeterBjuhr** [comment]
>Support for highlightedText - thanks to @PeterBjuhr . Confirmed not working on Ubuntu 12.04, still hoping it might work somewhere.

It works in Ubuntu 14.10. Here's an screenshot from the preferences:

![image](https://cloud.githubusercontent.com/assets/355116/6662534/ebde4720-cbbc-11e4-9a4a-2d2a19994ad7.png)

**Comment by PeterBjuhr** [comment]
>When a QIcon is added to a QAction, it can be used on buttons, menus and in the keyboard shortcuts list in the prefs, everywhere; you can't know beforehand. 

That sounds very reasonable. @proski 's patch works strangely well considering this. 

>Currently, the Snippets menu is redrawn on every display, but this might change in the future.

Does this mean that the actions is also recreated every time? That would indeed explain why the patch works.

**Comment by wbsoft** [comment]
Yes, the actions of the snippets menu are recreated every time on display of the snippet menu, because the symbol (icon) is requested every time.

But we should not depend on that behaviour, as the whole point of the `symbols` module is that the icons change their color dynamically. When a widget wants to paint the icon of an action, it requests the pixmap from the QIconEngine of the icon. (Currently based on app-wide default text color, but hopefully we can get it to work in a smarter way.)

I am now looking how a style paints icons and I see that the method `QStyle.generatedIconPixmap` is called to get the icon pixmap. That function can manipulate the pixmap before returning it. And it has access to style and palette information using the `QStyleOption`...

**Comment by PeterBjuhr** [comment]
>I am now looking how a style paints icons and I see that the method QStyle.generatedIconPixmap is called to get the icon pixmap. That function can manipulate the pixmap before returning it. And it has access to style and palette information using the `QStyleOption`...

I've also looked in that direction but without real success. I noticed that a new instance of `QStyleOption` was created.

**Comment by wbsoft** [comment]
To use the QStyle tricks I think we'd need to create a QStyle instance that proxies all method calls to the global QStyle object, except for the `generatedIconPixmap()`... We can also set a QStyle instance to the widget where we know that symbol icons are displayed. But then we also need to have a way to distinguish between normal icons and icons that need to be recolored, as we then only get the pixmap and not the icon name, etc.

To fix the bug in the current situation we could go with @proski's patch.

Or we create a fallback mode where the symbol icons are always black on a light background, drawn with friendly rounded corners. The mode can be activated by default if the global unity menubar is active, and we could even create a preference for it.

**Comment by wbsoft** [comment]
I see that I call `QApplication.style().generatedIconPixmap` already in the pixmap generating stage in `symbols/__init__.py`, but I think that's not necessary. I think that QStyle calls that method anyway with the pixmap...

**Comment by proski** [comment]
@PeterBjuhr My patch works because the menubar palette is specified where the icon is requested for the menu.

I'm glad the highlighting works for the snippets pane. I was too concerned with the menu to look elsewhere. Actually, the highlighting is triggered unnecessarily when using keyboard navigation in the snippets pane (Lubuntu 14.10, default theme, no appmenu), so it needs fixing.

I guess the best way to proceed would be to split highlighting into a separate patch and get it committed first.

Then we can make another patch to always take the palette for the correct widget. That would be a refinement to improve handling of complex themes, distinguish between `text()` and `foreground()` etc.

Then we can deal with Ubuntu appmenu. Is its activated, the palette for menu items would be taken from the menubar.

**Comment by proski** [comment]
Even the first stage (highlighting) cannot be done perfectly. The first picture is with focus. The second picture is without focus. In both cases, `pixmap()` is called with `'fonts_abc'`, `PyQt4.QtCore.QSize(16, 16)`, `QIcon.Selected`, `QIcon.Off`. We need to know something about the widget. Even if we remove the `pixmap` method and rely on `paint`, I don't see anything useful in QPainter.
![fonts-focus](https://cloud.githubusercontent.com/assets/1317472/6676369/9c244b8a-cbfd-11e4-9033-6d5aa907369b.png)
![fonts-nofocus](https://cloud.githubusercontent.com/assets/1317472/6676370/9e8b1a0c-cbfd-11e4-9c7c-fc10f8424256.png)

**Comment by wbsoft** [comment]
I have this idea: I originally designed the symbols to work on the quick insert panel. They are not meant to be used as icons in menus, although some snippets now use a symbol as icon.

I propose to create real icons for the built-in snippets that are in the snippets menu by default, (if they really need an icon).

Then we leave the symbol api implementation as it is currently.

We can make the symbols drawing nicer (in the `symbols.pixmap()` function) by optionally adding a background or a shadow, we could make that a preference.

**Comment by proski** [comment]
Yes, I tend to agree now. We tried to do something that is better done in Qt itself - an icon that changes color in sync with text. The existing functionality comes very close, but it still falls short of what is acceptable for a polished program. With 4 modes and 2 states, an icon can have 8 variants for a given pixel size, yet there is no indication of the changing focus, even though the text changes color. The icons for the Ubuntu appmenu don't request the highlight variant, even though they need it. Finding the widget by the icon would be a hack even if we manage to implement it, as it would go around the modes/states mechanism.

I'll' try to turn symbols used for the snippets into nice icons. Maybe something like tools-rhythm.svg or better yet list-add.svg, although doing that to an ellipsis could be tricky.

**Comment by wbsoft** [comment]
Ok, I think that will turn out nice.

**Comment by proski** [comment]
I have committed an icon for the fonts that is quite visible on any background because it uses multiple colors. I haven't been able to make good icons for the "Repeat" and "Last Note or Chord" yet.

Tango approach with double border makes the elements too thick. I don't think it's a good idea to draw abstract symbols like physical things. Symbols on white background don't look good if the menu background is light gray. Simply black doesn't look good on dark gray menus. I'm experimenting with black symbols with a white or light gray shadow around the perimeter, but I haven't made anything nice looking yet. Another approach would be light blue with a dark blue shadow.

"Repeat" could probably be simplified to something like this:
http://www.marathon-studios.com/unicode/U1D107/Musical_Symbol_Right_Repeat_Sign

In any case, this issue should not delay 2.18.1. It would fix much more important things.
